### PR TITLE
Check empty input data in binary search functions

### PR DESCRIPTION
**Pull request by SergeyKopienko**
Fixed errors in upper_bound, lower_bound, binary_search algorithms: unable to call these algorithms with empty source data.

### CODE REVIEW

File: test/general/sycl_iterator.pass.cpp [hunk 1]
```diff
@@ -3458,15 +3458,14 @@ int b[] = {0, 1, 1, 6, 6, 9};
...
-constexpr auto nc = sizeof(c) / sizeof(c[0]);
-constexpr auto nd = sizeof(d) / sizeof(d[0]);
+constexpr size_t na = sizeof(a) / sizeof(a[0]);
+constexpr size_t nb = sizeof(b) / sizeof(b[0]);
+constexpr size_t nc = sizeof(c) / sizeof(c[0]);
+constexpr size_t nd = sizeof(d) / sizeof(d[0]);
 
-template <typename Size>
-Size get_size(Size n)
+constexpr size_t get_size()
```
Comment (MikeDvorskiy) [inline test/general/sycl_iterator.pass.cpp:3466]: m.b. better to write 
`constexpr size_t n = na + nb + nc + nd;` ?

Comment (MikeDvorskiy) [inline test/general/sycl_iterator.pass.cpp:3466]: m.b. better to write 
`constexpr size_t n = na + nb + nc + nd;` ?

File: test/general/sycl_iterator.pass.cpp [hunk 2]
```diff
@@ -3477,8 +3476,11 @@ DEFINE_TEST(test_includes)
     void
     operator()(Policy&& exec, Iterator1 first1, Iterator1 last1, Iterator2 first2, Iterator2 last2, Size n)
     {
-        TestDataTransfer<UDTKind::eKeys, Size> host_keys(*this, get_size(n));
-        TestDataTransfer<UDTKind::eVals, Size> host_vals(*this, get_size(n));
+        if (n < get_size())
```
Comment (MikeDvorskiy) [inline test/general/sycl_iterator.pass.cpp:3479]: What does that comparison mean?

Comment (SergeyKopienko) [reply test/general/sycl_iterator.pass.cpp:3479]: This comparison mean that for this test some minimal buffer size required.

File: test/support/utils_sycl.h [hunk 3]
```diff
@@ -137,7 +137,7 @@ sycl::queue get_test_queue()
     return my_queue;
 }
 
-template <sycl::usm::alloc alloc_type, typename TestValueType, typename TestName>
+template <sycl::usm::alloc alloc_type, typename TestValueType, typename TestName, size_t kStartIndex = 1>
```
Comment (MikeDvorskiy) [inline test/support/utils_sycl.h:140]: Why do wee need such default value?
In my understanding each test should run algo on a range [0, max_n].

File: test/support/utils_sycl.h [hunk 4]
```diff
@@ -153,14 +153,25 @@ test1buffer()
...
-                                               n);
+            try
+            {
+                invoke_on_all_hetero_policies<0>()(create_test_obj<TestValueType, TestName>(test_base_data),
+                                                   inout1_offset_first, inout1_offset_first + n,
+                                                   n);
+            }
+            catch (const std::exception& exc)
+            {
+                ::std::cout << "Exception occurred in test1buffers function (USM) : " << exc.what() << std::endl;
```
Comment (MikeDvorskiy) [inline test/support/utils_sycl.h:169]: can we share that diagnostic (lines 169-171)  between six places in the code?

File: test/general/sycl_iterator.pass.cpp (hunk 1 above)
Comment (SergeyKopienko) [reply test/general/sycl_iterator.pass.cpp]: Fixed

File: test/support/utils_sycl.h (hunk 3 above)
Comment (SergeyKopienko) [reply test/support/utils_sycl.h:140]: I am tried to do this variant. But many changes required for this.
So I think that we can test from 0 only these tests, which are really required.

File: test/support/utils_sycl.h (hunk 4 above)
Comment (SergeyKopienko) [reply test/support/utils_sycl.h]: Fixed

File: test/parallel_api/algorithm/alg.sorting/alg.binary.search/binary.search/binary_search.pass.cpp [hunk 5]
```diff
@@ -135,15 +138,15 @@ main()
 
 #if TEST_DPCPP_BACKEND_PRESENT
     // Run tests for USM shared memory
-    test3buffers<sycl::usm::alloc::shared, test_binary_search<ValueType>>();
+    test3buffers<sycl::usm::alloc::shared, test_binary_search<ValueType>, 0>();
```
Comment (MikeDvorskiy) [inline test/parallel_api/algorithm/alg.sorting/alg.binary.search/binary.search/binary_search.pass.cpp:141]: What does zero mean?
(should me a short one-word comment, like ,` 0/*...*/>()`

**Review by MikeDvorskiy — APPROVED** [review]
LGTM

**Pull request merged**